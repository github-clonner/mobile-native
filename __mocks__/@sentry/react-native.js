export const captureMessage = jest.fn();
export default {};
