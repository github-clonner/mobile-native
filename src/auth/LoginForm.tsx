//@ts-nocheck
import React, { Component } from 'react';

import * as Animatable from 'react-native-animatable';

import {
  View,
  Text,
  LayoutAnimation,
  // TextInput,
} from 'react-native';

import Icon from 'react-native-vector-icons/Ionicons';

import authService from './AuthService';

import i18n from '../common/services/i18n.service';
import logService from '../common/services/log.service';
import Input from '../common/components/Input';
import Button from '../common/components/Button';
import ThemedStyles from '../styles/ThemedStyles';

type PropsType = {
  onLogin: Function;
  onForgot: Function;
};

/**
 * Login Form
 */
export default class LoginForm extends Component<PropsType> {
  /**
   * State
   */
  state = {
    username: '',
    password: '',
    msg: '',
    twoFactorToken: '',
    twoFactorCode: '',
    hidePassword: true,
    inProgress: false,
    showLanguages: false,
  };

  /**
   * Constructor
   */
  constructor(props: PropsType) {
    super(props);
    this.state.language = i18n.getCurrentLocale();
  }

  /**
   * Render
   */
  render() {
    const CS = ThemedStyles.style;

    const msg = this.state.msg ? (
      <Animatable.Text
        animation="bounceInLeft"
        useNativeDriver
        style={[CS.subTitleText, CS.colorSecondaryText, CS.textCenter]}
        testID="loginMsg">
        {this.state.msg}
      </Animatable.Text>
    ) : null;

    return (
      <View style={[CS.flexContainer, CS.marginTop6x]}>
        <Text style={[CS.titleText, CS.colorPrimaryText, CS.marginBottom2x]}>
          {i18n.t('auth.login')}
        </Text>
        {msg}
        <Input
          placeholder={i18n.t('auth.username')}
          onChangeText={this.setUsername}
          value={this.state.username}
          style={CS.marginBottom2x}
          testID="usernameInput"
        />
        <View>
          <Input
            placeholder={i18n.t('auth.password')}
            secureTextEntry={this.state.hidePassword}
            onChangeText={this.setPassword}
            value={this.state.password}
            style={CS.marginBottom2x}
            testID="userPasswordInput"
          />
          <Icon
            name={this.state.hidePassword ? 'md-eye' : 'md-eye-off'}
            size={25}
            onPress={this.toggleHidePassword}
            style={CS.inputIcon}
          />
        </View>
        <Button
          onPress={() => this.onLoginPress()}
          text={i18n.t('auth.login')}
          containerStyle={[CS.button, CS.fullWidth]}
          textStyle={CS.buttonText}
          key={1}
          loading={this.state.inProgress}
          loadingRight={true}
          disabled={this.state.inProgress}
          disabledStyle={CS.backgroundTransparent}
          testID="loginButton"
        />
        <View style={CS.marginTop4x}>
          <Text style={[CS.link, CS.fontL]} onPress={this.onForgotPress}>
            {i18n.t('auth.forgot')}
          </Text>
        </View>
      </View>
    );
  }

  /**
   * Show languages
   */
  showLanguages = () => {
    this.setState({ showLanguages: true });
  };

  /**
   * Language selected
   */
  languageSelected = (language) => {
    this.setState({ language, showLanguages: false });
    i18n.setLocale(language);
  };

  /**
   * Cancel language selection
   */
  cancel = () => {
    this.setState({ showLanguages: false });
  };

  /**
   * Set two factor
   * @param {string} value
   */
  setTwoFactor = (value) => {
    const twoFactorCode = String(value).trim();
    this.setState({ twoFactorCode });
  };

  /**
   * Set two factor
   * @param {string} value
   */
  setUsername = (value) => {
    const username = String(value).trim();
    this.setState({ username });
  };

  /**
   * Set two factor
   * @param {string} value
   */
  setPassword = (value) => {
    const password = String(value).trim();
    this.setState({ password });
  };

  /**
   * Set two factor
   * @param {string} value
   */
  toggleHidePassword = () => {
    this.setState({ hidePassword: !this.state.hidePassword });
  };

  /**
   * Handle forgot password
   */
  onForgotPress = () => {
    this.props.onForgot();
  };

  /**
   * On login press
   */
  onLoginPress() {
    this.setState({ msg: '', inProgress: true });
    // is two factor auth
    if (this.state.twoFactorToken) {
      authService
        .twoFactorAuth(this.state.twoFactorToken, this.state.twoFactorCode)
        .then((data) => {
          this.props.onLogin();
        })
        .catch((err) => {
          logService.exception('[LoginForm]', err);
        });
    } else {
      authService
        .login(this.state.username, this.state.password)
        .then((data) => {
          this.props.onLogin();
        })
        .catch((errJson) => {
          LayoutAnimation.configureNext(LayoutAnimation.Presets.spring);
          if (
            errJson.error === 'invalid_grant' ||
            errJson.error === 'invalid_client'
          ) {
            this.setState({
              msg: i18n.t('auth.invalidGrant'),
              inProgress: false,
            });
            return;
          }

          //TODO implement on backend and edit
          if (errJson.error === 'two_factor') {
            this.setState({
              twoFactorToken: errJson.message,
              inProgress: false,
            });
            return;
          }

          this.setState({
            msg: errJson.message || 'Unknown error',
            inProgress: false,
          });

          logService.exception('[LoginForm]', errJson);
        });
    }
  }
}
