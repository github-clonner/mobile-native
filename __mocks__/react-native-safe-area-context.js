import React from "react";
import { View } from "react-native";

export const SafeAreaView = jest.fn().mockReturnValue(function NavigationContainer(props) {return <View>{props.children}</View>;});