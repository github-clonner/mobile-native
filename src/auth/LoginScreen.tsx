//@ts-nocheck
import React, { Component } from 'react';

import {
  StyleSheet,
  View,
  Keyboard,
  Animated,
  Text,
  TouchableOpacity,
  ScrollView,
  Platform,
  KeyboardAvoidingView,
} from 'react-native';

import LoginForm from './LoginForm';
import logService from '../common/services/log.service';
import i18nService from '../common/services/i18n.service';
import sessionService from '../common/services/session.service';

import ThemedStyles from '../styles/ThemedStyles';
import { SafeAreaView } from 'react-native-safe-area-context';

import BannerInfo from '../topbar/BannerInfo';

const LOGO_HEIGHT = 80;
const LOGO_HEIGHT_SMALL = 40;

/**
 * Login screen
 */
export default class LoginScreen extends Component {
  state = {
    keyboard: false,
  };

  constructor(props) {
    super(props);

    this.logoHeight = new Animated.Value(LOGO_HEIGHT);
  }

  componentDidMount() {
    // Setting this here because if user register, then onboarding then logout and login again, will go to onboarding again
    sessionService.setInitialScreen('Tabs');
    this.keyboardWillShowSub = Keyboard.addListener(
      'keyboardDidShow',
      this.keyboardWillShow,
    );
    this.keyboardWillHideSub = Keyboard.addListener(
      'keyboardDidHide',
      this.keyboardWillHide,
    );
  }

  componentWillUnmount() {
    this.keyboardWillShowSub.remove();
    this.keyboardWillHideSub.remove();
  }

  keyboardWillShow = (event) => {
    Animated.timing(this.logoHeight, {
      duration: 500,
      toValue: LOGO_HEIGHT_SMALL,
    }).start();

    // this.setState({keyboard: true});
    // LayoutAnimation.configureNext(LayoutAnimation.Presets.spring);
  };

  keyboardWillHide = (event) => {
    Animated.timing(this.logoHeight, {
      duration: 500,
      toValue: LOGO_HEIGHT,
    }).start();

    // this.setState({keyboard: false});
    // LayoutAnimation.configureNext(LayoutAnimation.Presets.spring);
  };

  /**
   * Render
   */
  render() {
    const CS = ThemedStyles.style;

    return (
      <KeyboardAvoidingView
        style={[CS.flexColumnStretch, CS.backgroundPrimary]}
        behavior={Platform.OS == 'ios' ? 'padding' : null}>
        <SafeAreaView style={[styles.flex10]}>
          <BannerInfo logged={false} />
          <ScrollView style={CS.flexContainer} keyboardShouldPersistTaps={true}>
            <View style={[CS.paddingHorizontal4x, CS.flexColumnStretch]}>
              <Animated.Image
                resizeMode="contain"
                source={require('./../assets/logos/bulb.png')}
                style={[styles.bulb, { height: this.logoHeight }]}
              />
              <LoginForm
                onLogin={() => this.login()}
                onForgot={this.onPressForgot}
              />
            </View>
          </ScrollView>
        </SafeAreaView>
        <View
          style={[
            CS.paddingVertical2x,
            CS.backgroundSecondary,
            CS.mindsLayoutFooter,
          ]}>
          <TouchableOpacity
            onPress={this.onPressRegister}
            testID="registerButton">
            <View style={CS.flexColumnCentered}>
              <Text style={[CS.subTitleText, CS.colorSecondaryText]}>
                {i18nService.t('auth.haveAccount')}
              </Text>
              <Text style={[CS.titleText, CS.colorPrimaryText]}>
                {i18nService.t('auth.createChannel')}
              </Text>
            </View>
          </TouchableOpacity>
        </View>
      </KeyboardAvoidingView>
    );
  }

  /**
   * On press forgot
   */
  onPressForgot = () => {
    this.props.navigation.push('Forgot');
  };

  /**
   * On press register
   */
  onPressRegister = () => {
    this.props.navigation.push('Register');
  };

  /**
   * On login successful
   */
  login() {
    logService.info('user logged in');
  }
}

const styles = StyleSheet.create({
  flex10: {
    flex: 10,
  },
  bulb: {
    width: 40,
    height: 67,
    alignSelf: 'center',
    marginTop: 10,
  },
});
