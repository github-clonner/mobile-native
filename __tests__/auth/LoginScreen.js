import 'react-native';
import React from 'react';
import { Text, TouchableOpacity } from 'react-native';
import { shallow } from 'enzyme';
import LoginScreen from '../../src/auth/LoginScreen';
import LoginForm from '../../src/auth/LoginForm';
import { SafeAreaView } from 'react-native-safe-area-context';

jest.mock('../../src/auth/AuthService');

// jest.mock('../../src/auth/LoginForm', () => 'LoginForm');
jest.mock('../../src/auth/ForgotPassword', () => 'ForgotPassword');
jest.mock('react-native-safe-area-context');

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

describe('LoginScreen component', () => {
  beforeEach(() => {});

  it('should renders correctly', () => {
    const loginScreen = renderer.create(<LoginScreen />).toJSON();
    expect(loginScreen).toMatchSnapshot();
  });

  it('should shows login form component', async () => {
    const wrapper = shallow(<LoginScreen hopFeatures={true} />);

    // search login form
    let render = wrapper.dive();
    const loginForms = render.find(LoginForm);

    // should contain 1 login form
    expect(loginForms.length).toBe(1);
  });

  it('should shows forgot password component if the user press the button', async () => {
    const navigation = {
      push: jest.fn(),
    };

    const wrapper = shallow(<LoginScreen navigation={navigation} />);

    // search login form
    let render = wrapper.dive();
    const loginForms = render.find(LoginForm);

    // should contain 1 login form
    expect(loginForms.length).toBe(1);

    // simulate forgot password component button press
    loginForms
      .at(0)
      .props()
      .onForgot();

    expect(navigation.push).toBeCalled();
  });
});
