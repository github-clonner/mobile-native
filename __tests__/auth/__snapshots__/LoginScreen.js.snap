// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen component should renders correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Array [
        undefined,
        undefined,
      ],
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    style={
      Array [
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="registerButton"
    >
      <View>
        <Text
          style={
            Array [
              undefined,
              undefined,
            ]
          }
        >
          Don't have an account?
        </Text>
        <Text
          style={
            Array [
              undefined,
              undefined,
            ]
          }
        >
          Join Minds Now
        </Text>
      </View>
    </View>
  </View>
</View>
`;
